# 评估和清洗airbnb(旅行房屋租赁社区)数据

## 分析目标

清洗掉无用的租赁房屋数据，保留有价值的租赁房屋数据

## 简介

该数据集包含了2019年纽约市的Airbnb上线的房间情况。Airbnb是一个旅行房屋租赁社区，用户可通过网站或手机APP发布、搜索度假房屋租赁信息并在线预定。

### 变量含义：

- id：房间id
- name：房间名称
- host_id：房东id
- host_name：房东姓名
- neighbourhood_group：地区
- neighbourhood：街区
- latitude：纬度坐标
- longitude：经度坐标
- room_type：房间类型
- price：价格（美元）
- minimum_nights：最少预定夜晚数
- number_of_reviews：评论数量
- last_review：最新浏览
- reviews_per_month：每月浏览次数
- calculated_host_listings_count：房东挂出房子的数量
- availability_365：可预定房源的天数

## 读取数据

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv("airbnb_NYC_2019.csv")

In [3]:
original_data.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

## 评估数据

在这一部分，我将对在上一部分建立的original_data这个DataFrame所包含的数据进行评估。

评估主要从两个方面进行：结构和内容，即整齐度和干净度。数据的结构性问题指不符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”这三个标准，数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

### 评估数据整齐度

In [4]:
original_data.sample(10)

id                                               name    host_id  \
4224    2771126                    Large Loft in Brooklyn  w/ Deck    2333018   
5285    3819656                                SoHo/Village Studio   19644902   
44970  34485699                 Modern Two Queen Near Central Park  261761719   
17993  14094264                    Upper East side Cozy apartment.   32764610   
31725  24735722        Small Private Room # 1, single / Twin  bed.   43825074   
38109  30087377              Prime West Village Boutique Apartment   14379366   
23256  18835820             Quiet, Cozy UES Studio Near the Subway   52777892   
27296  21558111    Clean, cozy and bright room 4 mins from subway!   13400096   
24031  19376182         Peaceful Room in the heart of Clinton Hill   29243209   
48333  36205851  Entire Spacious Apartment in UWS near Central ...  140997060   

       host_name neighbourhood_group      neighbourhood  latitude  longitude  \
4224         Max            Brooklyn       Clinton Hill  40.68998  -73.96072   
5285     Russell           Manhattan  Greenwich Village  40.72774  -74.00004   
44970  WestHouse           Manhattan            Midtown  40.76479  -73.98307   
17993      Ilkay           Manhattan        East Harlem  40.79920  -73.93879   
31725      Masud            Brooklyn      Cypress Hills  40.68593  -73.87793   
38109      Perry           Manhattan       West Village  40.73630  -74.00494   
23256        Amy           Manhattan    Upper East Side  40.76844  -73.95341   
27296        Ron            Brooklyn      Crown Heights  40.67770  -73.94251   
24031      Sarah            Brooklyn       Clinton Hill  40.68618  -73.96179   
48333        Amy           Manhattan    Upper West Side  40.77824  -73.98325   

             room_type  price  minimum_nights  number_of_reviews last_review  \
4224   Entire home/apt    142              10                  8  2019-01-03   
5285   Entire home/apt    165               2                  2  2015-06-26   
44970     Private room    100               1                  0         NaN   
17993     Private room     75               1                  0         NaN   
31725     Private room     35              28                  9  2019-05-01   
38109  Entire home/apt    340              30                  2  2019-06-14   
23256  Entire home/apt     10               3                 10  2018-10-22   
27296     Private room     48              45                  3  2018-08-04   
24031     Private room     70               1                  6  2017-08-13   
48333  Entire home/apt    275               3                  0         NaN   

       reviews_per_month  calculated_host_listings_count  availability_365  
4224                0.13                               1               157  
5285                0.04                               1                 0  
44970                NaN                               6               299  
17993                NaN                               1                 0  
31725               0.63                               3               312  
38109               0.46                               1               100  
23256               0.39                               1                 0  
27296               0.15                               3                 0  
24031               0.24                               2                 0  
48333                NaN                               2                24

从抽样的10行数据数据来看，数据符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”，具体来看每行是关于某商品的一次交易，每列是交易相关的各个变量，因此不存在结构性问题，并且在抽象数据中可观察到有部分租赁房屋的可预定房源天数为0，这类数据对于租赁数据来说无意义，需要删除。

## 评估数据干净度

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

从输出结果来看，数据共有条观48895值，而name、host_name、last_review、reviews_per_month变量存在缺失值。

此外，last_review 的数据类型应为日期，应当进行数据格式转换。

### 评估缺失数据

在了解name存在缺失值后，根据条件提取出缺失观察值。

In [6]:
original_data[original_data["name"].isnull()]

id name   host_id host_name neighbourhood_group  \
2854    1615764  NaN   6676776     Peter           Manhattan   
3703    2232600  NaN  11395220      Anna           Manhattan   
5775    4209595  NaN  20700823     Jesse           Manhattan   
5975    4370230  NaN  22686810   Michaël           Manhattan   
6269    4581788  NaN  21600904     Lucie            Brooklyn   
6567    4756856  NaN   1832442  Carolina            Brooklyn   
6605    4774658  NaN  24625694      Josh           Manhattan   
8841    6782407  NaN  31147528  Huei-Yin            Brooklyn   
11963   9325951  NaN  33377685  Jonathan           Manhattan   
12824   9787590  NaN  50448556    Miguel           Manhattan   
13059   9885866  NaN  37306329  Juliette           Manhattan   
13401  10052289  NaN  49522403   Vanessa            Brooklyn   
15819  12797684  NaN  69715276       Yan           Manhattan   
16071  12988898  NaN  71552588    Andrea               Bronx   
18047  14135050  NaN  85288337      Jeff            Brooklyn   
28889  22275821  NaN  49662398  Kathleen            Brooklyn   

            neighbourhood  latitude  longitude        room_type  price  \
2854    Battery Park City  40.71239  -74.01620  Entire home/apt    400   
3703         East Village  40.73215  -73.98821  Entire home/apt    200   
5775    Greenwich Village  40.73473  -73.99244  Entire home/apt    225   
5975               Nolita  40.72046  -73.99550  Entire home/apt    215   
6269         Williamsburg  40.71370  -73.94378     Private room    150   
6567             Bushwick  40.70046  -73.92825     Private room     70   
6605   Washington Heights  40.85198  -73.93108     Private room     40   
8841         Williamsburg  40.71354  -73.93882     Private room     45   
11963      Hell's Kitchen  40.76436  -73.98573  Entire home/apt    190   
12824              Harlem  40.80316  -73.95189  Entire home/apt    300   
13059           Chinatown  40.71632  -73.99328     Private room     67   
13401         Brownsville  40.66409  -73.92314     Private room     50   
15819     Upper West Side  40.79843  -73.96404     Private room    100   
16071             Fordham  40.86032  -73.88493      Shared room    130   
18047  Bedford-Stuyvesant  40.69421  -73.93234     Private room     70   
28889            Bushwick  40.69546  -73.92741  Entire home/apt    110   

       minimum_nights  number_of_reviews last_review  reviews_per_month  \
2854             1000                  0         NaN                NaN   
3703                1                 28  2015-06-08               0.45   
5775                1                  1  2015-01-01               0.02   
5975                7                  5  2016-01-02               0.09   
6269                1                  0         NaN                NaN   
6567                1                  0         NaN                NaN   
6605                1                  0         NaN                NaN   
8841                1                  0         NaN                NaN   
11963               4                  1  2016-01-05               0.02   
12824               5                  0         NaN                NaN   
13059               4                  0         NaN                NaN   
13401               3                  3  2016-08-18               0.07   
15819               1                  0         NaN                NaN   
16071               1                  0         NaN                NaN   
18047               3                  0         NaN                NaN   
28889               4                  5  2018-08-13               0.27   

       calculated_host_listings_count  availability_365  
2854                                1               362  
3703                                1               341  
5775                                1                 0  
5975                                1                 0  
6269                                1                 0  
6567                                1    

有16条租赁数据缺失name变量值。


In [7]:
original_data[original_data["host_name"].isnull()]

id                                               name    host_id  \
360      100184                                          Bienvenue     526653   
2700    1449546                            Cozy Studio in Flatbush    7779204   
5745    4183989    SPRING in the City!! Zen-Style Tranquil Bedroom     919218   
6075    4446862                 Charming Room in Prospect Heights!   23077718   
6582    4763327                Luxurious, best location, spa inc'l   24576978   
8163    6292866                          Modern Quiet Gem Near All   32722063   
8257    6360224                    Sunny, Private room in Bushwick   33134899   
8852    6786181                       R&S Modern Spacious Hideaway   32722063   
9138    6992973                    1 Bedroom in Prime Williamsburg    5162530   
9817    7556587                               Sunny Room in Harlem   39608626   
14040  10709846                 Sunny, spacious room in Greenpoint    7822683   
14631  11553543                                  Cozy Room Astoria   26138712   
15174  12113879     Sunny, Large West Village 1 BR Near Everything    5300585   
19565  15648096              Spacious 2 bedroom close to Manhattan  100971588   
27777  21787094         Spacious and Bright Midtown East Apartment     415290   
27962  21867710  Amazing room in the centre on Manhatten! Welcome!  159156636   
28042  21901156  Best Deal! Lovely place in Manhattan! Time Squ...  159156636   
28274  21990808                 ***SLEEPING IN THE GOLDEN PLACE***  159156636   
30570  23660605    Private Bedroom w/ Balcony view of Central Park  177146433   
32193  25180572                     Beautiful Cozy Studio Apt. UWS  119609345   
38992  30468335            LARGE Prospect Lefferts Gardens Bedroom  228750026   

      host_name neighbourhood_group     neighbourhood  latitude  longitude  \
360         NaN              Queens    Queens Village  40.72413  -73.76133   
2700        NaN            Brooklyn          Flatbush  40.64965  -73.96154   
5745        NaN           Manhattan            Harlem  40.80606  -73.95061   
6075        NaN            Brooklyn     Crown Heights  40.67512  -73.96146   
6582        NaN            Brooklyn        Greenpoint  40.72035  -73.95355   
8163        NaN            Brooklyn     East Flatbush  40.65263  -73.93215   
8257        NaN            Brooklyn          Bushwick  40.70146  -73.92792   
8852        NaN            Brooklyn     East Flatbush  40.64345  -73.93643   
9138        NaN            Brooklyn      Williamsburg  40.71838  -73.95630   
9817        NaN           Manhattan            Harlem  40.82929  -73.94182   
14040       NaN            Brooklyn        Greenpoint  40.73539  -73.95838   
14631       NaN              Queens  Ditmars Steinway  40.77587  -73.91775   
15174       NaN           Manhattan           Chelsea  40.73949  -73.99801   
19565       NaN               Bronx        Highbridge  40.83844  -73.92489   
27777       NaN           Manhattan   Upper East Side  40.76436  -73.96836   
27962       NaN           Manhattan    Hell's Kitchen  40.75668  -73.99097   
28042       NaN           Manhattan    Hell's Kitchen  40.75656  -73.99063   
28274       NaN           Manhattan    Hell's Kitchen  40.75835  -73.99065   
30570       NaN           Manhattan       East Harlem  40.79766  -73.94824   
32193       NaN           Manhattan   Upper West Side  40.77701  -73.97667   
38992       NaN            Brooklyn          Flatbush  40.65152  -73.95271   

             room_type  price  minimum_nights  number_of_reviews last_review  \
360       Private room     50               1                 43  2019-07-08   
2700   Entire home/apt    100              30                 49  2017-01-02   
5745      Private room     86               3                 34  2019-05-23   
6075      Private room     50               1                  0         NaN   
6582   Entire home/apt    195               1                  1  2015-10-20   
8163   Entire home/apt     85    

有21条租赁数据缺失host_name变量值，从输出结果来看，这些缺失last_review的数据，reviews_per_month都为空值。为了验证猜想，我们增加筛选条件，看是否存在last_review变量缺失且reviews_per_month不为空值的数据

In [8]:
original_data[(original_data["last_review"].isnull()) & ~(original_data["reviews_per_month"].isnull())]

Empty DataFrame
Columns: [id, name, host_id, host_name, neighbourhood_group, neighbourhood, latitude, longitude, room_type, price, minimum_nights, number_of_reviews, last_review, reviews_per_month, calculated_host_listings_count, availability_365]
Index: []

In [9]:
original_data[(original_data["last_review"].isnull()) & (original_data["number_of_reviews"] != 0)]

Empty DataFrame
Columns: [id, name, host_id, host_name, neighbourhood_group, neighbourhood, latitude, longitude, room_type, price, minimum_nights, number_of_reviews, last_review, reviews_per_month, calculated_host_listings_count, availability_365]
Index: []

In [10]:
original_data[(original_data["reviews_per_month"].isnull()) & (original_data["number_of_reviews"] != 0)]

Empty DataFrame
Columns: [id, name, host_id, host_name, neighbourhood_group, neighbourhood, latitude, longitude, room_type, price, minimum_nights, number_of_reviews, last_review, reviews_per_month, calculated_host_listings_count, availability_365]
Index: []

筛选出来结果数量为0条，说明缺失last_review值的数据，同时reviews_per_month的值也为空，并且number_of_reviews的值都为0。
last_review表示最新浏览次数，reviews_per_month表示每月浏览次数，number_of_reviews表示评论数量。如果他们同时无效，表示该租赁信息无人浏览，这些房源并不受欢迎。

In [11]:
original_data[original_data.duplicated()]

Empty DataFrame
Columns: [id, name, host_id, host_name, neighbourhood_group, neighbourhood, latitude, longitude, room_type, price, minimum_nights, number_of_reviews, last_review, reviews_per_month, calculated_host_listings_count, availability_365]
Index: []

从输出结果看，租赁数据没有重复数据，无需处理。

## 评估不一致数据

In [12]:
original_data['neighbourhood_group'].value_counts()
original_data['neighbourhood_group'].value_counts()

neighbourhood_group
Manhattan        21661
Brooklyn         20104
Queens            5666
Bronx             1091
Staten Island      373
Name: count, dtype: int64

In [13]:
original_data['neighbourhood'].value_counts()

neighbourhood
Williamsburg          3920
Bedford-Stuyvesant    3714
Harlem                2658
Bushwick              2465
Upper West Side       1971
                      ... 
Fort Wadsworth           1
Richmondtown             1
New Dorp                 1
Rossville                1
Willowbrook              1
Name: count, Length: 221, dtype: int64

In [14]:
original_data["room_type"].value_counts()

room_type
Entire home/apt    25409
Private room       22326
Shared room         1160
Name: count, dtype: int64

从输出内容看，没有需要处理的不一致数据。

## 评估无效或错误数据

In [15]:
original_data.describe()

id       host_id      latitude     longitude         price  \
count  4.889500e+04  4.889500e+04  48895.000000  48895.000000  48895.000000   
mean   1.901714e+07  6.762001e+07     40.728949    -73.952170    152.720687   
std    1.098311e+07  7.861097e+07      0.054530      0.046157    240.154170   
min    2.539000e+03  2.438000e+03     40.499790    -74.244420      0.000000   
25%    9.471945e+06  7.822033e+06     40.690100    -73.983070     69.000000   
50%    1.967728e+07  3.079382e+07     40.723070    -73.955680    106.000000   
75%    2.915218e+07  1.074344e+08     40.763115    -73.936275    175.000000   
max    3.648724e+07  2.743213e+08     40.913060    -73.712990  10000.000000   

       minimum_nights  number_of_reviews  reviews_per_month  \
count    48895.000000       48895.000000       38843.000000   
mean         7.029962          23.274466           1.373221   
std         20.510550          44.550582           1.680442   
min          1.000000           0.000000           0.010000   
25%          1.000000           1.000000           0.190000   
50%          3.000000           5.000000           0.720000   
75%          5.000000          24.000000           2.020000   
max       1250.000000         629.000000          58.500000   

       calculated_host_listings_count  availability_365  
count                    48895.000000      48895.000000  
mean                         7.143982        112.781327  
std                         32.952519        131.622289  
min                          1.000000          0.000000  
25%                          1.000000          0.000000  
50%                          1.000000         45.000000  
75%                          2.000000        227.000000  
max                        327.000000        365.000000

从输出结果看，有price变量数据为0的情况，表示该房屋租赁价格为0;有availability_365变量数据为0的情况，表示该房屋暂时无法出租。这两类数据都为异常数据，price变量数据为0表示为错误数据，availability_365变量数据为0表示为无意义数据，两类异常数据都需要删除.

In [26]:
original_data[(original_data["availability_365"] == 0)]

id                                               name    host_id  \
4          5022   Entire Apt: Spacious Studio/Loft by central park       7192   
6          5121                                    BlissArtsSpace!       7356   
8          5203                 Cozy Clean Guest Room - Family Apt       7490   
14         6090                      West Village Nest - Superhost      11975   
20         7801                   Sweet and Spacious Brooklyn Loft      21207   
...         ...                                                ...        ...   
48550  36313048    Sunny room with private entrance in shared home   16883913   
48731  36410519    Sunlight charming apt. in the heart of Brooklyn  121384174   
48756  36419441                            Murray Hill Masterpiece  273824202   
48760  36420725  Sunnyside, Queens 15 Mins to Midtown Clean & C...   19990280   
48779  36424776  Queen size bedroom in two bed apt with great l...   13384464   

           host_name neighbourhood_group       neighbourhood  latitude  \
4              Laura           Manhattan         East Harlem  40.79851   
6              Garon            Brooklyn  Bedford-Stuyvesant  40.68688   
8          MaryEllen           Manhattan     Upper West Side  40.80178   
14             Alina           Manhattan        West Village  40.73530   
20             Chaya            Brooklyn        Williamsburg  40.71842   
...              ...                 ...                 ...       ...   
48550        Tiffany              Queens           Ridgewood  40.69919   
48731  Luciana Paula            Brooklyn          Park Slope  40.66716   
48756          David           Manhattan         Murray Hill  40.74404   
48760        Brandon              Queens           Sunnyside  40.74719   
48779         Hollie            Brooklyn  Bedford-Stuyvesant  40.69617   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
4      -73.94399  Entire home/apt     80              10                  9   
6      -73.95596     Private room     60              45                 49   
8      -73.96723     Private room     79               2                118   
14     -74.00525  Entire home/apt    120              90                 27   
20     -73.95718  Entire home/apt    299               3                  9   
...          ...              ...    ...             ...                ...   
48550  -73.89902     Private room     45               1                  0   
48731  -73.98101  Entire home/apt    111               8                  0   
48756  -73.97239  Entire home/apt    129               2                  0   
48760  -73.91919     Private room     46               1                  0   
48779  -73.94198     Private room     50               3                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
4      2018-11-19               0.10                               1   
6      2017-10-05               0.40                               1   
8      2017-07-21               0.99                               1   
14     2018-10-31               0.22                               1   
20     2011-12-28               0.07                               1   
...           ...                ...                             ...   
48550         NaN                NaN                               1   
48731         NaN                NaN                               1   
48756         NaN                NaN                               1   
48760         NaN                NaN                               1   
48779         NaN                NaN                               1   

       availability_365  
4                     0  
6                     0  
8                     0  
14                    0  
20                    0  
...                 ...  
48550                 0  
48731                 0  
48756                 0  
48760                 0  
48779                 0  

[17533 rows x 16 columns

In [27]:
original_data[(original_data["price"] == 0)]

id                                               name    host_id  \
23161  18750597  Huge Brooklyn Brownstone Living, Close to it all.    8993084   
25433  20333471      ★Hostel Style Room | Ideal Traveling Buddies★  131697576   
25634  20523843    MARTIAL LOFT 3: REDEMPTION (upstairs, 2nd room)   15787004   
25753  20608117                    Sunny, Quiet Room in Greenpoint    1641537   
25778  20624541      Modern apartment in the heart of Williamsburg   10132166   
25794  20639628  Spacious comfortable master bedroom with nice ...   86327101   
25795  20639792  Contemporary bedroom in brownstone with nice view   86327101   
25796  20639914       Cozy yet spacious private brownstone bedroom   86327101   
26259  20933849                              the best you can find   13709292   
26841  21291569  Coliving in Brooklyn! Modern design / Shared room  101970559   
26866  21304320             Best Coliving space ever! Shared room.  101970559   

          host_name neighbourhood_group       neighbourhood  latitude  \
23161      Kimberly            Brooklyn  Bedford-Stuyvesant  40.69023   
25433        Anisha               Bronx     East Morrisania  40.83296   
25634  Martial Loft            Brooklyn            Bushwick  40.69467   
25753        Lauren            Brooklyn          Greenpoint  40.72462   
25778       Aymeric            Brooklyn        Williamsburg  40.70838   
25794       Adeyemi            Brooklyn  Bedford-Stuyvesant  40.68173   
25795       Adeyemi            Brooklyn  Bedford-Stuyvesant  40.68279   
25796       Adeyemi            Brooklyn  Bedford-Stuyvesant  40.68258   
26259        Qiuchi           Manhattan         Murray Hill  40.75091   
26841        Sergii            Brooklyn            Bushwick  40.69211   
26866        Sergii            Brooklyn            Bushwick  40.69166   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
23161  -73.95428     Private room      0               4                  1   
25433  -73.88668     Private room      0               2                 55   
25634  -73.92433     Private room      0               2                 16   
25753  -73.94072     Private room      0               2                 12   
25778  -73.94645  Entire home/apt      0               5                  3   
25794  -73.91342     Private room      0               1                 93   
25795  -73.91170     Private room      0               1                 95   
25796  -73.91284     Private room      0               1                 95   
26259  -73.97597  Entire home/apt      0               3                  0   
26841  -73.90670      Shared room      0              30                  2   
26866  -73.90928      Shared room      0              30                  5   

      last_review  reviews_per_month  calculated_host_listings_count  \
23161  2018-01-06               0.05                               4   
25433  2019-06-24               2.56                               4   
25634  2019-05-18               0.71                               5   
25753  2017-10-27               0.53                               2   
25778  2018-01-02               0.15                               1   
25794  2019-06-15               4.28                               6   
25795  2019-06-21               4.37                               6   
25796  2019-06-23               4.35                               6   
26259         NaN                NaN                               1   
26841  2019-06-22               0.11                               6   
26866  2019-05-24               0.26                               6   

       availability_365  
23161                28  
25433               127  
25634                 0  
25753                 0  
25778                73  
25794               176  
25795               232  
25796               222  
26259                 0  
26841               333  
26866               139

# 清理数据

根据前面评估部分得到的结论，我们需要进行的数据清理包括：

把 last_review 的变量数据转化为日期时间

把 price为0的变量数据删除

把 availability_365为0的变量数据删除

为了区分开经过清理的数据和原始的数据，我们创建新的变量`cleaned_data`，让它为`original_data`复制出的副本。我们之后的清理步骤都将被运用在`cleaned_data`上。

In [18]:
cleaned_data = original_data.copy()
cleaned_data.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

将last_review的变量数据转化为日期时间

In [20]:
cleaned_data["last_review"] = pd.to_datetime(cleaned_data["last_review"])
cleaned_data["last_review"]

0       2018-10-19
1       2019-05-21
2              NaT
3       2019-07-05
4       2018-11-19
           ...    
48890          NaT
48891          NaT
48892          NaT
48893          NaT
48894          NaT
Name: last_review, Length: 48895, dtype: datetime64[ns]

把price变量值为0的观察值删除，并检查替换后price变量值为0的个数:

In [29]:
cleaned_data = cleaned_data[cleaned_data["price"] != 0]
len(cleaned_data[cleaned_data["price"] == 0])

0

把availability_365变量值为0的观察值删除，并检查替换后availability_365变量值为0的个数:

In [30]:
cleaned_data = cleaned_data[cleaned_data["availability_365"] != 0]
len(cleaned_data[cleaned_data["availability_365"] == 0])

0

### 保存清洗后的数据

完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为`airbnb_NYC_2019_cleaned.csv`。

In [32]:
cleaned_data.head()

id                                       name  host_id    host_name  \
0  2539         Clean & quiet apt home by the park     2787         John   
1  2595                      Skylit Midtown Castle     2845     Jennifer   
2  3647        THE VILLAGE OF HARLEM....NEW YORK !     4632    Elisabeth   
3  3831            Cozy Entire Floor of Brownstone     4869  LisaRoxanne   
5  5099  Large Cozy 1 BR Apartment In Midtown East     7322        Chris   

  neighbourhood_group neighbourhood  latitude  longitude        room_type  \
0            Brooklyn    Kensington  40.64749  -73.97237     Private room   
1           Manhattan       Midtown  40.75362  -73.98377  Entire home/apt   
2           Manhattan        Harlem  40.80902  -73.94190     Private room   
3            Brooklyn  Clinton Hill  40.68514  -73.95976  Entire home/apt   
5           Manhattan   Murray Hill  40.74767  -73.97500  Entire home/apt   

   price  minimum_nights  number_of_reviews last_review  reviews_per_month  \
0    149               1                  9  2018-10-19               0.21   
1    225               1                 45  2019-05-21               0.38   
2    150               3                  0         NaT                NaN   
3     89               1                270  2019-07-05               4.64   
5    200               3                 74  2019-06-22               0.59   

   calculated_host_listings_count  availability_365  
0                               6               365  
1                               2               355  
2                               1               365  
3                               1               194  
5                               1               129

In [33]:
cleaned_data.to_csv("airbnb_NYC_2019_cleaned.csv",index = False)

In [34]:
pd.read_csv("airbnb_NYC_2019_cleaned.csv").head()

id                                       name  host_id    host_name  \
0  2539         Clean & quiet apt home by the park     2787         John   
1  2595                      Skylit Midtown Castle     2845     Jennifer   
2  3647        THE VILLAGE OF HARLEM....NEW YORK !     4632    Elisabeth   
3  3831            Cozy Entire Floor of Brownstone     4869  LisaRoxanne   
4  5099  Large Cozy 1 BR Apartment In Midtown East     7322        Chris   

  neighbourhood_group neighbourhood  latitude  longitude        room_type  \
0            Brooklyn    Kensington  40.64749  -73.97237     Private room   
1           Manhattan       Midtown  40.75362  -73.98377  Entire home/apt   
2           Manhattan        Harlem  40.80902  -73.94190     Private room   
3            Brooklyn  Clinton Hill  40.68514  -73.95976  Entire home/apt   
4           Manhattan   Murray Hill  40.74767  -73.97500  Entire home/apt   

   price  minimum_nights  number_of_reviews last_review  reviews_per_month  \
0    149               1                  9  2018-10-19               0.21   
1    225               1                 45  2019-05-21               0.38   
2    150               3                  0         NaN                NaN   
3     89               1                270  2019-07-05               4.64   
4    200               3                 74  2019-06-22               0.59   

   calculated_host_listings_count  availability_365  
0                               6               365  
1                               2               355  
2                               1               365  
3                               1               194  
4                               1               129